src/actions/format.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Code formatting using Rustfmt - by default using statically-linked one or
+//! possibly running Rustfmt binary specified by the user.
+
+use std::env::temp_dir;
+use std::fs::File;
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
+
+use rand::{Rng, thread_rng};
+use log::{log, debug};
+use rustfmt_nightly::{Config, Session, Input};
+use serde_json;
+
+struct External<'a>(&'a Path, &'a Path);
+struct Internal;
+
+/// Specified which `rustfmt` to use.
+#[derive(Clone)]
+pub enum Rustfmt {
+    /// (Path to external `rustfmt`, cwd where it should be spawned at)
+    External(PathBuf, PathBuf),
+    /// Statically linked `rustfmt`
+    Internal
+}
+
+impl From<Option<(String, PathBuf)>> for Rustfmt {
+    fn from(value: Option<(String, PathBuf)>) -> Rustfmt {
+        match value {
+            Some((path, cwd)) => Rustfmt::External(PathBuf::from(path), cwd),
+            None => Rustfmt::Internal
+        }
+    }
+}
+
+
+pub trait Formatter {
```

Comment:
This trait seems like over-abstraction. Since we have the enum why not just use an inherent function on the enum and call different functions. Alternatively, keep the trait but dump the enum, since they are basically doing the same thing.

Reply:
It somehow felt cleaner doing this this way than writing two bare `format_internal(String, Config) -> ..` and `format_external(PathBuf, PathBuf, String, Config) -> ...` to be called from inherent function but you're right that there's no benefit apart from being able to implement the same function signature for different variants; I'll change that.